tree.go

```diff
@@ -1266,3 +1266,155 @@ func ToDot(root VerkleNode) string {
+			}
+			ret = append(ret, sn)
+		case *LeafNode:
+			c1Bytes := compressedPoints[0]
+			c2Bytes := compressedPoints[1]
+			sn := SerializedNode{
+				Node:            n,
+				CommitmentBytes: compressedPoints[0],
+				SerializedBytes: n.serializeWithCompressedCommitments(c1Bytes, c2Bytes),
+			}
+			ret = append(ret, sn)
+		}
+		compressedPoints = compressedPoints[pointToCompressPerNodeCount[i]:]
+	}
+
+	return ret, nil
+}
+
+func (n *InternalNode) collectNonHashedNodes(list []VerkleNode) []VerkleNode {
+	list = append(list, n)
+	for _, child := range n.children {
+		switch childNode := child.(type) {
+		case *LeafNode:
+			list = append(list, childNode)
+		case *InternalNode:
+			list = childNode.collectNonHashedNodes(list)
+		}
+	}
+	return list
+}
+
+func (n *InternalNode) serializeWithCompressedChildren(compressedChildren [][32]byte) []byte {
+	var (
+		bitlist, hashlist [NodeWidth / 8]byte
+		nhashed           int // number of children who are hashed nodes
+	)
+	for i, c := range n.children {
+		if _, ok := c.(Empty); !ok {
+			setBit(bitlist[:], i)
+			if _, ok := c.(*HashedNode); ok {
+				setBit(hashlist[:], i)
+				nhashed++
+			}
+		}
+	}
+
+	ret := make([]byte, nodeTypeSize+bitlistSize+(len(compressedChildren)+nhashed)*SerializedPointCompressedSize)
+
+	children := ret[internalNodeChildrenOffset:internalNodeChildrenOffset]
+	consumed := 0
+	for i := 0; i < NodeWidth; i++ {
+		if bit(bitlist[:], i) {
+			if bit(hashlist[:], i) {
+				children = append(children, n.children[i].(*HashedNode).commitment...)
+			} else {
+				children = append(children, compressedChildren[consumed][:]...)
+				consumed++
+			}
+		}
+	}
+
+	// Store in ret the serialized result
+	ret[nodeTypeOffset] = internalRLPType
+	copy(ret[internalBitlistOffset:], bitlist[:])
+	// Note that children were already appended in ret through the children slice.
+
+	return ret
+}
+
+func (n *LeafNode) serializeWithCompressedCommitments(c1Bytes [32]byte, c2Bytes [32]byte) []byte {
+	// Empty value in LeafNode used for padding.
+	var emptyValue [LeafValueSize]byte
+
+	// Create bitlist and store in children LeafValueSize (padded) values.
+	children := make([]byte, 0, NodeWidth*LeafValueSize)
+	var bitlist [bitlistSize]byte
+	for i, v := range n.values {
+		if v != nil {
+			setBit(bitlist[:], i)
+			children = append(children, v...)
+			if padding := emptyValue[:LeafValueSize-len(v)]; len(padding) != 0 {
+				children = append(children, padding...)
+			}
+		}
+	}
+
+	// Create the serialization.
+	baseSize := nodeTypeSize + StemSize + bitlistSize + 2*SerializedPointCompressedSize
+	result := make([]byte, baseSize, baseSize+4*32) // Extra pre-allocated capacity for 4 values.
+	result[0] = leafRLPType
+	copy(result[leafSteamOffset:], n.stem[:StemSize])
+	copy(result[leafBitlistOffset:], bitlist[:])
+	copy(result[leafC1CommitmentOffset:], c1Bytes[:])
+	copy(result[leafC2CommitmentOffset:], c2Bytes[:])
+	result = append(result, children...)
+
+	return result
+}
```

Comment:
A final comment about this PR.
I kept the existing `Flush(...)` API despite we [moved away from using it in go-etheruem](https://github.com/gballet/go-ethereum/pull/172).

Maybe it makes sense to keep it if for other uses cases it might be useful, but [I'd say probably isn't](https://github.com/gballet/go-ethereum/pull/172#discussion_r1124345678). In any case, if we later decide to remove that API, I think we might need to do a decent refactor in tests. (That's fine, but if that's the direction we're going let's plan for that in a separate PR)

Reply:
Yes, let's slowly convert the tests over time, but as long as it is used in tests, `Flush` is useful. It also needs to be replaced in other branches, like the new replay/convert branches I hope to release in the next few days.

Reply:
> Yes, let's slowly convert the tests over time, but as long as it is used in tests, Flush is useful. 
:+1: 

>  It also needs to be replaced in other branches, like the new replay/convert branches I hope to release in the next few days.

Yeah, I did that already in https://github.com/gballet/go-ethereum/pull/172 that you'll review next.